packages/langium-cli/src/generator/highlighting/monarch-generator.ts

```diff
@@ -0,0 +1,476 @@
+    return prettyPrint(monarchGrammar);
+}
+
+/**
+ * Gets Monarch tokenizer states from a Langium grammar
+ * @param grammar Langium grammar to source tokenizer states from
+ * @returns Array of tokenizer states
+ */
+function getTokenizerStates(grammar: langium.Grammar): State[] {
+
+    // initial state, name is arbitrary, just needs to come first
+    const initialState: State = {
+        name: 'initial',
+        rules: getTerminalRules(grammar)
+    };
+
+    const whitespaceState: State = {
+        name: 'whitespace',
+        rules: getWhitespaceRules(grammar)
+    };
+
+    const commentState: State = {
+        name: 'comment',
+        rules: getCommentRules(grammar)
+    };
+
+    // order the following additional rules, to prevent
+    // comment sequences being classified as symbols
+
+    // add include for the whitespace state
+    initialState.rules.push(whitespaceState);
+
+    // add operator & symbol case handling
+    initialState.rules.push({
+        regex: '@symbols',
+        action: [
+            {
+                guard: '@operators',
+                action: { token: 'operator' }
+            },
+            // by default, leave the symbol alone
+            {
+                guard: '@default',
+                action: { token: '' }
+            }
+        ]
+    });
+
+    return [
+        initialState,
+        whitespaceState,
+        commentState
+    ];
+}
+
+/**
+ * Pretty prints a monarch grammar into it's concrete form, suitable for writing to a file
+ * @param monarchGrammar Grammar to pretty print
+ * @returns Monarch grammar in concrete form, suitable for writing to a file
+ */
+function prettyPrint(monarchGrammar: MonarchGrammar): string {
+    const name = monarchGrammar.languageDefinition.name;
+    return ([
+        `// Monarch syntax highlighting for the ${name} language.`,
+        `export const language${name} = {\n`,
+
+        // add language definitions
+        prettyPrintLangDef(monarchGrammar.languageDefinition),
+
+        // add folding
+        '\tfolding: {',
+        '\t\tmarkers: {',
+        '\t\t\tstart: new RegExp(\'^\\s*//\\s*#?region\\b\'),',
+        '\t\t\tend: new RegExp(\'^\\s*//\\s*#?endregion\\b\')',
+        '\t\t}',
+        '\t},\n',
+
+        // add tokenizer parts, simple state machine groupings
+        prettyPrintTokenizer(monarchGrammar.tokenizer),
+
+        '};'
+    ].join('\n')).replaceAll(/\t/g, '    ');
+}
+
+/**
+ * Generates an entry for a language definition, given a name (token category) and values
+ * @param name Category of language definition to add
+ * @param values Values to add under the given category
+ * @param fmt Formatter to keep things indented
+ * @returns A string of this language def entry, for use in a monarch file
+ */
+function genLanguageDefEntry(name: string, values: string[], fmt: Formatter): string {
+    return [
+        `${name}: [`,
+        '\t' + values.map(v => `'${v}'`).join(','),
```

Comment:
Suggestion: Please use the `GeneratorNode` infrastructure like the other generator files (except for the `TextMate` stuff since it just uses `JSON.stringify`)